src/main/java/picard/sam/SamToFastq.java

```diff
@@ -286,21 +260,70 @@ protected int doWork() {
                 // Create this writer on-the-fly; if we find no second-of-pair reads, don't bother making a writer (or delegating,
                 // if we're interleaving).
                 final Lazy<FastqWriter> lazySecondOfPairWriter = new Lazy<>(() -> INTERLEAVE ? firstOfPairWriter : factory.newWriter(makeReadGroupFile(rg, "_2")));
-                writerMap.put(rg, new FastqWriters(firstOfPairWriter, lazySecondOfPairWriter, firstOfPairWriter));
+
+                writerMap.put(rg, new FastqWriters(firstOfPairWriter, lazySecondOfPairWriter, firstOfPairWriter, additionalWriters));
             }
         }
         return writerMap;
     }
 
+    protected void initializeAdditionalWriters() {
+    }
+
+    protected Map<SAMReadGroupRecord, List<FastqWriter>> generateAdditionalWriters(List<SAMReadGroupRecord> readGroups,
+                                                                                   FastqWriterFactory factory) {
+        return Collections.emptyMap();
+    }
+
+    private void handleRecord(final SAMRecord currentRecord, final Map<SAMReadGroupRecord, FastqWriters> writers,
+                              final Map<String, SAMRecord> firstSeenMates) {
+        if (currentRecord.isSecondaryOrSupplementary() && !INCLUDE_NON_PRIMARY_ALIGNMENTS)
+            return;
+
+        // Skip non-PF reads as necessary
+        if (currentRecord.getReadFailsVendorQualityCheckFlag() && !INCLUDE_NON_PF_READS)
+            return;
+
+        final FastqWriters fq = writers.get(currentRecord.getReadGroup());
+        SAMRecord read1 = null;
+        SAMRecord read2 = null;
+        if (currentRecord.getReadPairedFlag()) {
+            final String currentReadName = currentRecord.getReadName();
+            final SAMRecord firstRecord = firstSeenMates.remove(currentReadName);
+            if (firstRecord == null) {
+                firstSeenMates.put(currentReadName, currentRecord);
+            } else {
+                assertPairedMates(firstRecord, currentRecord);
+
+                read1 = currentRecord.getFirstOfPairFlag() ? currentRecord : firstRecord;
+                read2 = currentRecord.getFirstOfPairFlag() ? firstRecord : currentRecord;
+                writeRecord(read1, 1, fq.getFirstOfPair(), READ1_TRIM, READ1_MAX_BASES_TO_WRITE);
+                final FastqWriter secondOfPairWriter = fq.getSecondOfPair();
+                if (secondOfPairWriter == null) {
+                    throw new PicardException("Input contains paired reads but no SECOND_END_FASTQ specified.");
+                }
+                writeRecord(read2, 2, secondOfPairWriter, READ2_TRIM, READ2_MAX_BASES_TO_WRITE);
+            }
+        } else {
+            writeRecord(currentRecord, null, fq.getUnpaired(), READ1_TRIM, READ1_MAX_BASES_TO_WRITE);
+        }
+
+        handleAdditionalRecords(currentRecord, fq.additionalWriters, read1, read2);

```

Comment:
As with `initializeAdditionalWriters()` above, I think it would be clearer to have the subclass override `handleRecord()` and call `super.handleRecord()` to execute the code in the base class.